feat: auth pages improvement for mobile and desktop platform

hyj1204 opened the pull request:
<!---
Thank you for submitting a pull request to AppFlowy. The team will dedicate their best efforts to reviewing and approving your pull request. If you have any questions about the project or feedback for us, please join our [Discord](https://discord.gg/wdjWUXXhtw).
-->

<!---
If your pull request adds a new feature, please drag and drop a video into this section to showcase what you've done! If not, you may delete this section.
-->
##  Organize the code
- Change the term ‘Welcome’  to ‘Wrokspace’, and `WelcomeScreen` becomes `WorkspaceStartPage`. 
- Move all the authorization screens under one folder(appflowy_flutter/lib/user/presentation/screens)
 -  Add static const `routeName` in every authorization screen, and keep it the same as the screen name. In this way to prevent misspelling the route name. 
 > [!NOTE]
> Every time, when we need a route name, just simply use `**Screen.routeName` to navigate to `**Screen` 
 - Separate mobile theme and desktop theme data
 
##  Auth Screens
For all the auth screens, they shared the same workflow(same bloc and state), the only difference between mobile and desktop is their UIs. Thus, for every auth screen, it has two versions of UIs, one is for mobile, and another is for desktop. 
For example, we have `MobileSignInScreen` and `DesktopSignInScreen` under `SignInScreen`. Both `MobileSignInScreen` and `DesktopSignInScreen` use the widgets like `SignInAnonymousButton` and 'ThirdPartySignInButtons' to achieve the same functionalities with different UIs(`isMobile` to control which UI will apply)
The following screens have been separated into mobile and desktop:
- SignInScreen
- WorkspaceStartScreen
- MobileHomeScreen and DesktopHomeScreen


I haven't added any test code yet. I plan to add it once the authentication workflow is more stable.


### DesktopSignInScreen
<img width="792" alt="image" src="https://github.com/AppFlowy-IO/AppFlowy/assets/14248245/cd27bdf3-2c47-4a33-a93c-9945cdab13b0">

### MobileSignInScreen
<img width="383" alt="Pasted Graphic 2" src="https://github.com/AppFlowy-IO/AppFlowy/assets/14248245/0fe95ce2-99cd-4e17-a0a4-a9ef9fae84d1">

### Google login workflow

https://github.com/AppFlowy-IO/AppFlowy/assets/14248245/fcf8fde5-82f4-487d-9aa7-e4c1565563e4













<!---
List at least one issue here that this PR addresses. If it fixes the issue, please use the [fixes](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests) keyword to close the issue. For example:
fixes https://github.com/AppFlowy-IO/AppFlowy/pull/2106
-->

---

<!---
Before you mark this PR ready for review, run through this checklist!
-->

#### PR Checklist

- [X] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [X] I've listed at least one issue that this PR fixes in the description above.
- [ ] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [X] All existing tests are passing.

appflowy commented on frontend/appflowy_flutter/lib/user/presentation/router.dart:
```diff
@@ -138,26 +151,43 @@ class SplashRoute {
-          userProfile,
-          workspaceSetting,
-          key: ValueKey(userProfile.id),
+    if (Platform.isAndroid || Platform.isIOS) {
```
Would it be better if we replace all existing Platform.isAndroid || Platform.isIOS to a single function?

appflowy commented on frontend/appflowy_flutter/lib/user/presentation/router.dart:
```diff
@@ -138,26 +151,43 @@ class SplashRoute {
-      ),
-    );
+        // pop up all the pages until [SplashScreen]
+        (route) => route.isFirst,
```
IIRC, you can retrieve the name of the router and check whether it corresponds to the splash screen

appflowy commented on frontend/appflowy_flutter/lib/user/presentation/screens/sign_in_screen/widgets/sign_in_anonymous_button.dart:
```diff
@@ -0,0 +1,106 @@
+  final bool isMobile;
+
+  /// Used in DesktopSignInScreen and MobileSignInScreen
+  const SignInAnonymousButton({
```
I edited the same button here. You might need to rebase after this PR get merged: https://github.com/AppFlowy-IO/AppFlowy/pull/3250

LucasXu0 commented on frontend/appflowy_flutter/lib/user/presentation/screens/sign_in_screen/desktop_sign_in_screen.dart:
```diff
@@ -0,0 +1,253 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+
+class DesktopSignInScreen extends StatefulWidget {
```
StatelessWidget?

LucasXu0 commented on frontend/appflowy_flutter/lib/user/presentation/screens/sign_in_screen/desktop_sign_in_screen.dart:
```diff
@@ -0,0 +1,253 @@
+        ),
+        Padding(
+          padding: EdgeInsets.symmetric(horizontal: 10),
+          child: FlowyText.regular('OR'),
```
i18n.

LucasXu0 commented on frontend/appflowy_flutter/lib/user/presentation/screens/sign_in_screen/mobile_sign_in_screen.dart:
```diff
@@ -0,0 +1,98 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+
+class MobileSignInScreen extends StatefulWidget {
```
StatelessWidget?

LucasXu0 commented on frontend/appflowy_flutter/lib/user/presentation/screens/sign_in_screen/mobile_sign_in_screen.dart:
```diff
@@ -0,0 +1,98 @@
+                  const VSpace(16),
+                  // TODO(yijing): confirm the subtitle before release app
+                  Text(
+                    'You are in charge of your data and customizations.',
```
i18n.

LucasXu0 commented on frontend/appflowy_flutter/lib/user/presentation/screens/sign_in_screen/mobile_sign_in_screen.dart:
```diff
@@ -0,0 +1,98 @@
+                  const VSpace(spacing * 2),
+                  // Welcome to
+                  Text(
+                    welcomeString.substring(0, welcomeString.length - 8),
```
The `welcomeString` is a localization string. It's unreasonable to calculate its length minus 8.

LucasXu0 commented on frontend/appflowy_flutter/lib/user/presentation/screens/sign_in_screen/sign_in_screen.dart:
```diff
@@ -0,0 +1,81 @@
+  }
+}
+
+void handleOpenWorkspaceError(BuildContext context, FlowyError error) {
```
I noticed that you have defined it in 'handle_open_workspace_error.dart'. Please use the same function and remove the duplicated one.

<img width="462" alt="Screenshot 2023-08-24 at 16 11 04" src="https://github.com/AppFlowy-IO/AppFlowy/assets/11863087/fc037681-9b12-42a1-b60a-04893ab33514">

LucasXu0 commented on frontend/appflowy_flutter/lib/user/presentation/screens/sign_in_screen/widgets/sign_in_anonymous_button.dart:
```diff
@@ -0,0 +1,85 @@
+  final bool isMobile;
+
+  /// Used in DesktopSignInScreen and MobileSignInScreen
+  const SignInAnonymousButton({
+    Key? key,
+    required this.isMobile,
+  }) : super(key: key);
```
```suggestion
  const SignInAnonymousButton({
    super.key,
    required this.isMobile,
  });
```

LucasXu0 commented on frontend/appflowy_flutter/lib/user/presentation/screens/sign_in_screen/widgets/third_party_sign_in_buttons.dart:
```diff
@@ -0,0 +1,168 @@
+  final Alignment contentAlignment;
+
+  /// Used in DesktopSignInScreen and MobileSignInScreen
+  const ThirdPartySignInButtons({
+    Key? key,
+    required this.isMobile,
+    this.contentAlignment = Alignment.center,
+  }) : super(key: key);
```
use super.key instead.

hyj1204 replied:
Why do we need to change it?

LucasXu0 replied:
The new syntax is much shorter and cleaner.

LucasXu0 commented on frontend/appflowy_flutter/lib/user/presentation/screens/splash_screen.dart:
```diff
@@ -44,6 +56,10 @@ class SplashScreen extends StatelessWidget {
       child: Scaffold(
         body: BlocListener<SplashBloc, SplashState>(
           listener: (context, state) {
+            Log.debug(
+              'SplashScreen -> SplashState/AuthState: ${state.auth.toString()}',
+            );
```
remove it.

LucasXu0 commented on frontend/appflowy_flutter/lib/user/presentation/screens/workspace_start_screen/desktop_workspace_start_screen.dart:
```diff
@@ -0,0 +1,92 @@
+import 'package:appflowy/generated/locale_keys.g.dart';
```
afaik, the desktop workspace page should be deprecated. cc @appflowy

hyj1204 replied:
From the previous code, I thought the workspace start screen was to let users choose which workspace they are going to or create a new workspace. What is the correct workflow after user login? @appflowy

hyj1204 replied:
I did some updates [here](https://github.com/AppFlowy-IO/AppFlowy/pull/3217#discussion_r1304940076), and I left `WorkspaceStartScreen` anyway, we may need it to ask the user which workspace they want to open.

LucasXu0 commented on frontend/appflowy_flutter/lib/user/presentation/screens/workspace_start_screen/util/create_workspace.dart:
```diff
@@ -0,0 +1,14 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+
+void createWorkspace(BuildContext context) {
+  context.read<WorkspaceBloc>().add(
+        WorkspaceEvent.createWorkspace(
+          LocaleKeys.workspace_hint.tr(),
+          "",
+        ),
+      );
+}
```
Not every context can access the workspace bloc, so defining it as a global function can be risky.

LucasXu0 commented on frontend/appflowy_flutter/lib/user/presentation/screens/workspace_start_screen/util/pop_to_workspace.dart:
```diff
@@ -0,0 +1,10 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+
+void popToWorkspace(BuildContext context, WorkspacePB workspace) {
```
same as here.

hyj1204 commented on frontend/appflowy_flutter/lib/mobile/presentation/mobile_home_page.dart:
```diff
@@ -0,0 +1,46 @@
+    this.workspaceSetting, {
+    super.key,
+  });
+  static const routeName = "/MobileHomeScreen";
```
The reason why I define the route name within every screen is to build the connection between them.

For example, when navigating to `SignInScreen`, its route name will be `SignInScreen.routeName`. Thus, we can always know which page it leads to by clicking `SignInScreen` on `SignInScreen.routeName`. In any place we put the route name, we can immediately know which page it leads to.  

The approach you suggested has a downside; wrong/misspelling of the route name can make it difficult to locate the correct screen. For example, if we change the screen name to MobileHomeFirstScreen but forget to change the route name, it will cause unnecessary confusion in finding the target screen through its route name.

LucasXu0 replied:
It would be difficult to list all the supported router names if you define them separately.

hyj1204 replied:
You can use 
```
@override
  Widget build(BuildContext context) {
  return MaterialApp(
    routes: {
      HomeScreen.routeName: (context) => HomeScreen(),
      DetailScreen.routeName: (context) => DetailScreen(),
    },
  );
}

```
Instead of
```
@override
  Widget build(BuildContext context) {
  return MaterialApp(
    routes: {
      '/': (context) => HomeScreen(),
      '/details': (context) => DetailScreen(),
    },
  );
}
```

hyj1204 commented on frontend/appflowy_flutter/integration_test/util/auth_operation.dart:
```diff
@@ -7,15 +7,15 @@ import 'base.dart';
 extension AppFlowyAuthTest on WidgetTester {
   Future<void> tapGoogleLoginInButton() async {
-    await tapButton(find.byType(GoogleSignUpButton));
+    await tapButton(find.byKey(const Key('signInWithGoogleButton')));
```
find.text doesn't work well in these highly customized widgets. Another reason I do not prefer to use it is texts could change over time(even for the localkeys names), and some texts could be used in multiple places, so when we debug, it brings more work to find the target widget we are testing for. 
By using the Key we can easily find the target widget by searching the key name, and it usually won't change over time as long as it is the same widget.

hyj1204 commented on frontend/appflowy_flutter/lib/user/presentation/helpers/handle_open_workspace_error.dart:
```diff
@@ -0,0 +1,25 @@
+
+void handleOpenWorkspaceError(BuildContext context, FlowyError error) {
+  if (error.code == ErrorCode.WorkspaceDataNotSync) {
+    final userFolder = UserFolderPB.fromBuffer(error.payload);
+    getIt<AuthRouter>().pushWorkspaceErrorScreen(context, userFolder, error);
```
@appflowy 
Could you help add some comments here? It looks like we got the user folder from the error message and we need to show this folder info in the workspace error screen, but I am not sure if that folder is the local one or the cloud one.

hyj1204 commented on frontend/appflowy_flutter/lib/user/presentation/screens/sign_in_screen/sign_in_screen.dart:
```diff
@@ -53,7 +59,7 @@ class SignInScreen extends StatelessWidget {
           router.pushEncryptionScreen(context, user);
         } else {
-          router.pushWorkspaceStartScreen(context, user);
+          router.pushHomeOrWorkspaceStartScreen(context, user);
```
@LucasXu0 @appflowy 
I noticed I used the wrong route during the naming changes. It should be the same as the main branch now.

Xazin commented on frontend/appflowy_flutter/lib/mobile/presentation/mobile_home_page.dart:
```diff
@@ -0,0 +1,47 @@
+
+class MobileHomeScreen extends StatelessWidget {
+  const MobileHomeScreen({
+    Key? key,
```
We should prefer `super.key` in all new or refactored classes.

Xazin commented on frontend/appflowy_flutter/lib/mobile/presentation/mobile_home_page.dart:
```diff
@@ -0,0 +1,47 @@
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: const Text("MobileHomeScreen"),
```
Title

Xazin commented on frontend/appflowy_flutter/lib/user/presentation/router.dart:
```diff
@@ -1,76 +1,91 @@
-          workspaceSetting,
-          key: ValueKey(profile.id),
+    if (PlatformExtension.isMobile) {
+      Navigator.pushAndRemoveUntil<void>(
```
I prefer 
```suggestion
      Navigator.of(context).pushAndRemoveUntil(
```

hyj1204 commented on frontend/appflowy_flutter/lib/user/presentation/screens/sign_in_screen/desktop_sign_in_screen.dart:
```diff
@@ -0,0 +1,245 @@
+            ),
+            const VSpace(20),
+            // loading status
+            ...isLoading
+                ? [
+                    const VSpace(indicatorMinHeight),
+                    const LinearProgressIndicator(
+                      value: null,
+                      minHeight: indicatorMinHeight,
+                    ),
+                  ]
+                : [
+                    const VSpace(indicatorMinHeight * 2.0)
+                  ], // add the same space when there's no loading status.
+            // ConstrainedBox(
```
It seems `value` should show the current progress of 'login with Google'. However, I didn't see any place where we can get it, so the LinearProgressIndicator shows a regular loop animation now.
I also changed the code like this

```
    // loading status
            const VSpace(indicatorMinHeight),
            isLoading
                ? const LinearProgressIndicator(
                    value: null,
                    minHeight: indicatorMinHeight,
                  )
                : const VSpace(indicatorMinHeight),
```

appflowy commented on frontend/appflowy_flutter/lib/user/presentation/screens/sign_in_screen/desktop_sign_in_screen.dart:
```diff
@@ -0,0 +1,242 @@
+            const _OrDivider(),
+            const VSpace(10),
+            const ThirdPartySignInButtons(
+              isMobile: false,
```
Would it be better to remove the `isMobile` parameter and encapsulate the platform logic within the `SignInAnonymousButton`?

hyj1204 replied:
Good idea! I just updated the code in both `SignInAnonymousButton` and `ThirdPartySignInButtons` for it.

appflowy commented:
I've updated the logic for opening a workspace that has errors.

hyj1204 commented:
I added discord and GitHub login buttons.
Light and dark mode in desktop: I used the different logo styles for different modes for a better visible appearance. 
<img width="813" alt="Pasted Graphic 14" src="https://github.com/AppFlowy-IO/AppFlowy/assets/14248245/1e25bd55-b20f-4fb0-898f-14e80cd8ea88">
<img width="803" alt="Pasted Graphic 12" src="https://github.com/AppFlowy-IO/AppFlowy/assets/14248245/0a618eb2-a7e8-4c04-b9c4-6bf66e6ae713">

Light mode in mobile: We haven't add dark mode for mobile yet
<img width="378" alt="Pasted Graphic 13" src="https://github.com/AppFlowy-IO/AppFlowy/assets/14248245/a91643af-d61c-4357-a829-89a83b5cc72e">

annieappflowy commented:
Logos should be of the same size and aligned vertically

hyj1204 commented:
I removed the padding inside google SVG file and updated the layout of `ThirdPartySignInButtons`.  In order to align all the labels vertically in a relatively centered position to the button, I use a fixed-width container(grey area) to wrap the icon(align to the right), then use another container to align the label to the left. Due to its unique alignment, I remove the `contentAlignment` from ThirdPartySignInButtons. 

<img width="394" alt="image" src="https://github.com/AppFlowy-IO/AppFlowy/assets/14248245/ec74e552-0277-4114-9620-0b6a17a275c0">

Final Screenshot
---
In Desktop: 
<img width="445" alt="Pasted Graphic 24" src="https://github.com/AppFlowy-IO/AppFlowy/assets/14248245/e0a7e436-4e19-43be-bdd4-3298cbe32c1e">

---
In Mobile
<img width="329" alt="Pasted Graphic 22" src="https://github.com/AppFlowy-IO/AppFlowy/assets/14248245/1a1d8bec-f4d0-4a33-ba7b-5082a541c0ef">
